src/dev/code_coverage/ingest_coverage/transforms.js

```
@@ -0,0 +1,165 @@
+const setTotal = x => obj => (obj.isTotal = x);
+const mutateTrue = setTotal(true);
+const mutateFalse = setTotal(false);
+
+const root = urlBase => ts => testRunnerType =>
+  `${urlBase}/${ts}/${testRunnerType.toLowerCase()}-combined`;
+
+const prokForTotalsIndex = mutateTrue => urlRoot => obj =>
+  right(obj)
+    .map(mutateTrue)
+    .map(always(`${urlRoot}/index.html`))
+    .fold(noop, id);
+
+const prokForCoverageIndex = root => mutateFalse => urlRoot => obj => siteUrl =>
+  right(siteUrl)
+    .map(x => {
+      mutateFalse(obj);
+      return x;
+    })
+    .map(x => {
+      return splitOnRoot(root)(x);
+    })
+    .map(x => `${urlRoot}${x}.html`)
+    .fold(noop, id);
+
+export const staticSite = urlBase => obj => {
+  const { staticSiteUrl, testRunnerType, COVERAGE_INGESTION_KIBANA_ROOT } = obj;
+  const ts = obj['@timestamp'];
+  const urlRoot = root(urlBase)(ts)(testRunnerType);
+  const prokTotal = prokForTotalsIndex(mutateTrue)(urlRoot);
+  const prokCoverage = prokForCoverageIndex(COVERAGE_INGESTION_KIBANA_ROOT)(mutateFalse)(urlRoot)(
+    obj
+  );
+  const prokForBoth = always(maybeTotal(staticSiteUrl).fold(always(prokTotal(obj)), prokCoverage));
+
+  return { ...obj, staticSiteUrl: prokForBoth() };
+};
+
+export const coveredFilePath = obj => {
+  const { staticSiteUrl, COVERAGE_INGESTION_KIBANA_ROOT } = obj;
+
+  const withoutCoveredFilePath = always(obj);
+  const leadingSlashRe = /^\//;
+  const maybeDropLeadingSlash = x => (leadingSlashRe.test(x) ? right(x) : left(x));
+  const dropLeadingSlash = x => x.replace(leadingSlashRe, '');
+  const dropRoot = root => x =>
+    maybeDropLeadingSlash(splitOnRoot(root)(x)).fold(id, dropLeadingSlash);
+  return maybeTotal(staticSiteUrl)
+    .map(dropRoot(COVERAGE_INGESTION_KIBANA_ROOT))
+    .fold(withoutCoveredFilePath, coveredFilePath => ({ ...obj, coveredFilePath }));
+};
+
+export const ciRunUrl = obj => {
+  const ciRunUrl = process.env.CI_RUN_URL || 'CI RUN URL NOT PROVIDED';
```

Should probably throw an error instead of setting a error message to the expected URL variable.

---

Well, it's not absolutely needed, yet anyway.  You prefer to throw?

---

One reason I'd prefer not to throw as this datum is not mission critical.  It takes over 2 hours to get here in ci.  We'd rather not throw

---

That's fine, then we should find another way not to fail if it's not set. We should never be setting a URL variable to something other than a URL, it's just asking for trouble.

---

Ahh, I we could truncate to just the ci url, minus the build number! Whatcha think?

---

I think that's misusing a mechanism to control flow which it's not meant to do. Can we just avoid triggering the code coverage ingest if CI_RUN_URL is not set?

---

Asking my boss now

---

Yep, I think if we don't have a CI_RUN_URL (like a local run) we just don't ingest that field.

---

I'll code that